From 93faa8e186c7ceaa809e05fda6f92e56ded9d858 Mon Sep 17 00:00:00 2001
From: Julian Bouzas <julian.bouzas@collabora.com>
Date: Mon, 20 Jul 2026 09:54:48 -0400
Subject: [PATCH] hwdb: classify PlayStation controller audio as controller
 form-factor

Add Sony PlayStation controller entries to 70-sound-card.hwdb so that their
ALSA sound devices are tagged with SOUND_FORM_FACTOR=controller:
 - DualSense (054c:0ce6)
 - DualSense Edge (054c:0df2)
 - DualShock 4 CUH-ZCT1x (054c:05c4)
 - DualShock 4 CUH-ZCT2x (054c:09cc)

These controllers expose USB audio but are neither headsets nor speakers.
Pinning them in the hwdb ensures they are identified correctly before any
fallback matching occurs.
---
 hwdb.d/70-sound-card.hwdb | 13 +++++++++++++
 hwdb.d/parse_hwdb.py      | 16 +++++++++++++++-
 2 files changed, 28 insertions(+), 1 deletion(-)

diff --git a/hwdb.d/70-sound-card.hwdb b/hwdb.d/70-sound-card.hwdb
index 762993dbf07..a121c0b0575 100644
--- a/hwdb.d/70-sound-card.hwdb
+++ b/hwdb.d/70-sound-card.hwdb
@@ -37,6 +37,19 @@ usb:v045Ep02F9*
 usb:v045Ep091E*
  SOUND_FORM_FACTOR=headset
 
+###########################################################
+# Sony
+###########################################################
+# DualSense (PS5)
+usb:v054Cp0CE6*
+# DualSense Edge (PS5)
+usb:v054Cp0DF2*
+# DualShock 4 [CUH-ZCT1x]
+usb:v054Cp05C4*
+# DualShock 4 [CUH-ZCT2x]
+usb:v054Cp09CC*
+ SOUND_FORM_FACTOR=controller
+
 ###########################################################
 # Steelseries
 ###########################################################
diff --git a/hwdb.d/parse_hwdb.py b/hwdb.d/parse_hwdb.py
index 9a599e735c3..4e867968a23 100755
--- a/hwdb.d/parse_hwdb.py
+++ b/hwdb.d/parse_hwdb.py
@@ -199,7 +199,21 @@ def property_grammar():
              ('ID_TAG_MASTER_OF_SEAT', Literal('1')),
              ('ID_INFRARED_CAMERA', Or((Literal('0'), Literal('1')))),
              ('ID_CAMERA_DIRECTION', Or(('front', 'rear'))),
-             ('SOUND_FORM_FACTOR', Or(('internal', 'webcam', 'speaker', 'headphone', 'headset', 'handset', 'microphone'))),
+             (
+                'SOUND_FORM_FACTOR',
+                Or(
+                    (
+                        'internal',
+                        'webcam',
+                        'speaker',
+                        'headphone',
+                        'headset',
+                        'handset',
+                        'microphone',
+                        'controller',
+                    )
+                ),
+            ),
             )
     fixed_props = [Literal(name)('NAME') - Suppress('=') - val('VALUE')
                    for name, val in props]
-- 
2.50.1

